topology/probes/vpp/vpp.go

```diff
@@ -0,0 +1,410 @@
+	tr.AddMetadata("Speed", int64(intf.LinkSpeed))
+	state := interfaceUpDown(intf.AdminUpDown)
+	tr.AddMetadata("State", state)
+	if state != "DOWN" {
+		tr.AddMetadata("Duplex", interfaceDuplex(intf.LinkDuplex))
+	}
+	tr.AddMetadata("VrfID", p.getInterfaceVrfID(ch, intf.SwIfIndex))
+
+	return node
+}
+
+func interfaceNeedUpdate(i1, i2 *interfaces.SwInterfaceDetails) bool {
+	if (i1.LinkMtu != i2.LinkMtu) ||
+		(i1.LinkSpeed != i2.LinkSpeed) ||
+		(i1.LinkDuplex != i2.LinkDuplex) {
+		return true
+	}
+	return false
+}
+
+func (p *Probe) eventAddInterface(ch api.Channel, intf *interfaces.SwInterfaceDetails) {
+	p.createOrUpdateInterface(ch, intf)
+}
+
+func (p *Probe) eventDelInterface(node *graph.Node) {
+	p.graph.Lock()
+	defer p.graph.Unlock()
+	if err := p.graph.DelNode(node); err != nil {
+		logging.GetLogger().Error(err)
+	}
+}
+
+func (p *Probe) interfaceEventsEnableDisable(ch api.Channel, enable bool) {
+	ed := uint32(0)
+	if enable {
+		ed = uint32(1)
+	}
+	req := &interfaces.WantInterfaceEvents{
+		EnableDisable: ed,
+		PID:           uint32(os.Getpid()),
+	}
+	msg := &interfaces.WantInterfaceEventsReply{}
+	err := ch.SendRequest(req).ReceiveReply(msg)
+	if err != nil {
+		logging.GetLogger().Error(err)
+	}
+}
+
+func (p *Probe) interfacesEvents() {
+	defer p.wg.Done()
+
+	ch, err := p.conn.NewAPIChannel()
+	if err != nil {
+		logging.GetLogger().Error("API channel error: ", err)
+		return
+	}
+
+	sub, err := ch.SubscribeNotification(p.notifChan, &interfaces.SwInterfaceEvent{})
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return
+	}
+
+	p.interfaceEventsEnableDisable(ch, true)
+
+	for atomic.LoadInt64(&p.state) == common.RunningState {
+		notif := <-p.notifChan
+		if notif == nil {
+			break
+		}
+		msg := notif.(*interfaces.SwInterfaceEvent)
+
+		node := p.getInterface(msg.SwIfIndex)
+		p.graph.RLock()
+		name, _ := node.GetField("Name")
+		p.graph.RLock()
+		if msg.Deleted > 0 {
+			logging.GetLogger().Debugf("Delete interface %v idx %d", name, msg.SwIfIndex)
+			p.eventDelInterface(node)
+			continue
+		}
+		logging.GetLogger().Debugf("ChangeState interface %v idx %d updown %d", name, msg.SwIfIndex, msg.AdminUpDown)
+		p.graph.Lock()
+		node.Metadata.SetField("State", interfaceUpDown(msg.AdminUpDown))
+		p.graph.Unlock()
+	}
+
+	p.interfaceEventsEnableDisable(ch, false)
+
+	sub.Unsubscribe()
+	ch.Close()
+}
+
+func (p *Probe) interfacesPolling() {
+	defer p.wg.Done()
+
+	ch, err := p.conn.NewAPIChannel()
+	if err != nil {
+		logging.GetLogger().Error("API channel error: ", err)
+		return
+	}
+
+	for atomic.LoadInt64(&p.state) == common.RunningState {
+		foundInterfaces := make(map[uint32]struct{})
+		needUpdate := make(map[uint32]struct{})
+
+		req := &interfaces.SwInterfaceDump{}
+		reqCtx := ch.SendMultiRequest(req)
+		for {
+			msg := &interfaces.SwInterfaceDetails{}
+			stop, err := reqCtx.ReceiveReply(msg)
+			if stop {
+				break
+			}
+			if err != nil {
+				logging.GetLogger().Error(err)
+				goto nextEvents
```

Comment:
you can remove this goto as it used only once

```
sleep
continue
```

Reply:
I can't continue I'm inside a for/while(1) loop
in case of error I would like to send a new request